src/core/components/pin-set.js

```diff
@@ -0,0 +1,275 @@
+          if (err) {
+            callback.fired = true
+            return callback(err)
+          }
+          dag.put(child, (err) => {
+            if (callback.fired) { return }
+            if (err) {
+              callback.fired = true
+              return callback(err)
+            }
+            logInternalKey(child.multihash)
+            rootLinks[this.h] = new DAGLink(
+              '', child.size, child.multihash
+            )
+            _done++
+            if (_done === _subcalls) {
+              // all finished
+              DAGNode.create(rootData, rootLinks, (err, rootNode) => {
+                if (err) { return callback(err) }
+                return callback(null, rootNode)
+              })
+            }
+          })
+        }
+        _subcalls += Object.keys(hashed).length
+        for (let h in hashed) {
+          if (hashed.hasOwnProperty(h)) {
+            pinSet.storeItems(
+              hashed[h],
+              logInternalKey,
+              storeItemsCb.bind({h: h}),
+              _subcalls,
+              _done
+            )
+          }
+        }
+      }
+    },
+
+    loadSet: (rootNode, name, logInternalKey, callback) => {
+      // callback (err, keys)
+      const link = rootNode.links.filter(l => l.name === name).pop()
+      if (!link) { return callback('No link found with name ' + name) }
+      logInternalKey(link.multihash)
+      dag.get(new CID(link.multihash), (err, res) => {
+        if (err) { return callback(err) }
+        const keys = []
+        const walkerFn = (link) => {
+          keys.push(link.multihash)
+        }
+        pinSet.walkItems(res.value, walkerFn, logInternalKey, (err) => {
+          if (err) { return callback(err) }
+          return callback(null, keys)
+        })
+      })
+    },
+
+    walkItems: (node, walkerFn, logInternalKey, callback) => {
+      // callback (err)
+      const h = readHeader(node)
+      if (h.err) { return callback(h.err) }
+      const fanout = h.header.fanout
+      let subwalkCount = 0
+      let finishedCount = 0
+
+      const walkCb = (err) => {
+        if (err) { return callback(err) }
+        finishedCount++
+        if (subwalkCount === finishedCount) {
+          return callback()
+        }
+      }
+
+      for (let i = 0; i < node.links.length; i++) {
+        const link = node.links[i]
+        if (i >= fanout) {
+          // item link
+          walkerFn(link, i, h.data)
+        } else {
+          // fanout link
+          logInternalKey(link.multihash)
+          if (!emptyKey.equals(link.multihash)) {
+            subwalkCount++
+            dag.get(new CID(link.multihash), (err, res) => {
+              if (err) { return callback(err) }
+              pinSet.walkItems(
+                res.value, walkerFn, logInternalKey, walkCb
+              )
+            })
+          }
+        }
+      }
+      if (!subwalkCount) {
+        return callback()
+      }
+    }
+  }
+  return pinSet
+}
```

Comment:
There is a ton of good work! 👌🏽

At the current moment, seems a bit complex, might be good to factor it out into separate files and individually test each.